### PR TITLE
[BYOC][TRT] Add Support for Dynamic batch size for Dense, Conv, Conv_transpose etc.

### PR DESCRIPTION
**Pull request by codeislife99**
This PR adds support for dynamic batch size for dense, conv , conv_transpose and ops surrounding them

### CODE REVIEW

**Comment by codeislife99** [comment]
cc: @anijain2305 @trevor-m @zhiics

File: python/tvm/relay/tensorrt.py [hunk 1]
```diff
@@ -439,25 +445,47 @@ def reshape_whitelist_fn(attrs, args):  # pylint: disable=unused-variable
...
-            # Resolve -1.
-            for i, value in enumerate(new_shape):
-                if value == -1:
-                    new_shape[i] = original_volume // np.prod([x for x in new_shape if x != -1])
-            # Remove batch dimension and see if volumes match
-            if shape[0] != new_shape[0]:
-                print("reshape: can't modify batch dimension.")
-                return False
+
+            dynamic_reshape = False
```
Comment (trevor-m) [inline python/tvm/relay/tensorrt.py]: This can be reduced to `dynamic_reshape = any([isinstance(x, tvm.tir.expr.Any) for x in shape])`

File: tests/python/relay/test_tensorrt.py [hunk 2]
```diff
@@ -849,10 +849,128 @@ def load_graph():
...
 
 
+def test_tensorrt_dynamic_batch():
+    if should_skip():
+        return
+
+    batches_to_test = [1, 1, 2, 3, 1, 3, 2]
+    x_shape = (relay.Any(), 1, 8, 8)
+    x_data = np.ones([max(batches_to_test)] + list(x_shape)[1:]).astype("float32")
+    # print(x_data.shape)
```
Comment (trevor-m) [inline tests/python/relay/test_tensorrt.py]: Remove comment

File: src/relay/transforms/annotate_target.cc [hunk 3]
```diff
@@ -77,7 +77,13 @@ class AnnotateTargetRewriter : public ExprRewriter {
         compiler_ends.push_back(call->args[0]);
       } else if (op_expr_to_target_.find(arg) != op_expr_to_target_.end()) {
         arg_target = op_expr_to_target_[arg];
-        compiler_ends.push_back(InsertAnnotation(arg, arg_target, make_end_op));
+        // If an argument is a call node and has no argument, then it should be tensor ops such as
```
Comment (anijain2305) [inline src/relay/transforms/annotate_target.cc]: We should not merge this to dev. We should directly bring the relevant PR from Apache TVM

Comment (anijain2305) [reply src/relay/transforms/annotate_target.cc]: What do you think @trevor-m

File: tests/python/relay/test_pass_annotate_target.py [hunk 4]
```diff
@@ -510,6 +510,91 @@ def after():
     assert tvm.ir.structural_equal(expected, result)
 
 
+def test_if_free_vars():
```
Comment (anijain2305) [inline tests/python/relay/test_pass_annotate_target.py]: Use the Apache TVM PR for this

File: python/tvm/relay/tensorrt.py [hunk 5]
```diff
@@ -220,24 +220,30 @@ def register_tensorrt_annotations(trt_version, use_implicit_batch=True):
     # _register_external_op_helper("slice_like")
 
     def add_whitelist_fn(attrs, args):  # pylint: disable=unused-variable
+        static_add = True
         for arg in args:
-            if not arg.checked_type.shape:
+            if not arg.checked_type.shape[1:]:
```
Comment (anijain2305) [inline python/tvm/relay/tensorrt.py]: Will this cause failure if shape is empty - scalar?

File: python/tvm/relay/tensorrt.py [hunk 6]
```diff
@@ -220,24 +220,30 @@ def register_tensorrt_annotations(trt_version, use_implicit_batch=True):
     # _register_external_op_helper("slice_like")
 
     def add_whitelist_fn(attrs, args):  # pylint: disable=unused-variable
+        static_add = True
```
Comment (trevor-m) [inline python/tvm/relay/tensorrt.py]: Instead of having separate code paths to handle static vs dynamic add, lets process the shape first to replace Any with -1:

`shapes= [[int(x) if not isinstance(x, tvm.tir.expr.Any) else -1 for x in args.checked_type.shape] for arg in args]`

Then in all further code we only need to look at `shapes`.

File: python/tvm/relay/tensorrt.py [hunk 7]
```diff
@@ -439,25 +445,47 @@ def reshape_whitelist_fn(attrs, args):  # pylint: disable=unused-variable
...
+                # Make sure that the batch dim is unmodified.
+                if int(new_shape[0]) < 0:
+                    for shape_val, new_shape_val in enumerate(shape[1:], new_shape[1:]):
+                        if not (
+                            isinstance(shape_val, int)
+                            and isinstance(new_shape_val, int)
+                            and int(shape_val) == int(new_shape_val)
+                        ):
+                            return False
+                return True
```
Comment (trevor-m) [inline python/tvm/relay/tensorrt.py:473]: If new_shape[0] is > 0, we still need to check it as best as we can.

File: tests/python/relay/test_tensorrt.py [hunk 8]
```diff
@@ -849,10 +849,128 @@ def load_graph():
...
+    for i in range(len(batches_to_test)):
+        assert_result_matches(result_dict[(i, True)], result_dict[(i, False)])
+
+
+def test_tensorrt_dynamic_dense():
+    if should_skip():
+        return
+    batches_to_test = [1, 1, 2, 3, 1, 3, 2]
+    x_shape = (relay.Any(), 25)
+    # x_shape = (2, 25)
```
Comment (trevor-m) [inline tests/python/relay/test_tensorrt.py]: remove comment

File: tests/python/relay/test_tensorrt.py [hunk 9]
```diff
@@ -849,10 +849,128 @@ def load_graph():
...
+
+
 if __name__ == "__main__":
     test_tensorrt_ops()
     test_tensorrt_simple()
     test_tensorrt_simple_cpu_io()
     test_tensorrt_not_compatible()
     test_tensorrt_integration()
     test_tensorrt_serialize()
+    test_tensorrt_dynamic_batch()
```
Comment (trevor-m) [inline tests/python/relay/test_tensorrt.py:913]: We probably don't need all 4 of these tests. I think just conv2d is good enough.

**Review by trevor-m — CHANGES_REQUESTED** [review]

File: python/tvm/relay/tensorrt.py (hunk 5 above)
Comment (codeislife99) [reply python/tvm/relay/tensorrt.py]: Hmm, let me add a test case for this, I was thinking that if its a scalar shape would be an empty list rather than None

Comment (trevor-m) [reply python/tvm/relay/tensorrt.py]: Should this line be `if len(arg.checked_type.shape) < 2`

Comment (codeislife99) [reply python/tvm/relay/tensorrt.py]: What about tensors of dimension 1 ?

Comment (trevor-m) [reply python/tvm/relay/tensorrt.py]: If its 1-d, then its just the batch dimension. So when it goes to TRT with implicit batching. it would become a scalar. Not sure if we have handled that situation in TRT - but since we have already been letting them through we can do `len(arg.checked_type.shape) < 1` for now

File: python/tvm/relay/analysis/analysis.py [hunk 10]
```diff
@@ -319,6 +319,22 @@ def get_total_mac_number(expr):
     return _ffi_api.GetTotalMacNumber(expr)
 
 
+def is_compute_heavy_graph(expr):
```
Comment (trevor-m) [inline python/tvm/relay/analysis/analysis.py]: I think it might be better to have this implemented in python and localized to tensorrt.py script. I don't know if this visitor is general enough to be merged to upstream like this

File: src/relay/transforms/annotate_target.cc [hunk 11]
```diff
@@ -114,7 +114,6 @@ class AnnotateTargetRewriter : public ExprRewriter {
      * new expression.
      *
      *  This function checks for expr existence within the map and inserts the annotation
-     *  Further, it propagates the target to the new expression and returns it
```
Comment (trevor-m) [inline src/relay/transforms/annotate_target.cc]: remove this change

**Review by trevor-m — CHANGES_REQUESTED** [review]

File: tests/python/relay/test_pass_annotate_target.py (hunk 4 above)
Comment (codeislife99) [reply tests/python/relay/test_pass_annotate_target.py]: Yes, done

File: python/tvm/relay/analysis/analysis.py (hunk 10 above)
Comment (anijain2305) [reply python/tvm/relay/analysis/analysis.py]: Agreed. Lets try to keep the changes local to tensorrt.py. Anything outside needs to go to Apache TVM

Comment (codeislife99) [reply python/tvm/relay/analysis/analysis.py]: I see, okay , I will make the changes

File: python/tvm/relay/tensorrt.py (hunk 1 above)
Comment (codeislife99) [reply python/tvm/relay/tensorrt.py]: Done

File: python/tvm/relay/tensorrt.py (hunk 5 above)
Comment (codeislife99) [reply python/tvm/relay/tensorrt.py]: Done

File: python/tvm/relay/tensorrt.py (hunk 7 above)
Comment (codeislife99) [reply python/tvm/relay/tensorrt.py:473]: Done..

File: tests/python/relay/test_tensorrt.py (hunk 9 above)
Comment (codeislife99) [reply tests/python/relay/test_tensorrt.py:913]: dynamic_batch()mand dynamic_batch_conv() go through different passes, so although we technically dont need dynamic_batch() , do you think its a good idea to just keep it since it goes through a different pass than dynamic_batch_conv()

**Comment by codeislife99** [comment]
cc : @anijain2305 @trevor-m Please re-review.

File: python/tvm/relay/tensorrt.py (hunk 7 above)
Comment (trevor-m) [reply python/tvm/relay/tensorrt.py:473]: Thanks, looks good

File: python/tvm/relay/tensorrt.py [hunk 12]
```diff
@@ -718,7 +749,41 @@ def visit_call(self, call):
         return super().visit_call(call)
 
 
-def PruneSubgraphs(mod, compiler="tensorrt", use_implicit_batch=True, prune_no_macs=False):
+class ComputeHeavyGraph(ExprMutator):
```
Comment (trevor-m) [inline python/tvm/relay/tensorrt.py]: Should be ExprVisitor since it won't modify. Maybe should rename to `IsComputeHeavyGraph` also

File: python/tvm/relay/analysis/analysis.py (hunk 10 above)
Comment (codeislife99) [reply python/tvm/relay/analysis/analysis.py]: Done

File: python/tvm/relay/tensorrt.py [hunk 13]
```diff
@@ -738,7 +803,7 @@ def PruneSubgraphs(mod, compiler="tensorrt", use_implicit_batch=True, prune_no_m
         Which mode we plan to use for TensorRT. Will be used to determine which subgraphs are
         valid. In implicit batch mode, all inputs to a subgraph must have the same batch size.
 
-    prune_no_macs : bool
+    prune_compute_heavy_subgraphs : bool
```
Comment (trevor-m) [inline python/tvm/relay/tensorrt.py]: I don't think we need to change this name

File: tests/python/relay/test_tensorrt.py (hunk 9 above)
Comment (trevor-m) [reply tests/python/relay/test_tensorrt.py:913]: Sure that is fine

File: python/tvm/relay/tensorrt.py [hunk 14]
```diff
@@ -718,7 +749,41 @@ def visit_call(self, call):
...
+    Visits the Graph recursively and checks if it contains compute heavy ops like convolutions and
+    its transpose, dense and batch mat-mul.
+    """
+
+    def __init__(self):
+        ExprMutator.__init__(self)
+        self.is_compute_heavy = False
+
+    def visit_call(self, call):
+        heavy_ops = set(
```
Comment (anijain2305) [inline python/tvm/relay/tensorrt.py:763]: compute_intensive might be better name than compute_heavy

Comment (codeislife99) [reply python/tvm/relay/tensorrt.py:763]: Done

File: python/tvm/relay/tensorrt.py (hunk 13 above)
Comment (codeislife99) [reply python/tvm/relay/tensorrt.py]: Reverted

File: python/tvm/relay/tensorrt.py (hunk 12 above)
Comment (codeislife99) [reply python/tvm/relay/tensorrt.py]: Done.

**Review by trevor-m — APPROVED** [review]
